postgres - add indexes to foreign keys

lynzt opened the pull request:
## Description

We don't have indexes on foreign keys in our postgres db. This adds indexes to our foreign keys.

## Reviewer Notes

I think I caught them all, but worth a double check?

## Setup

`make db_dev_e2e_populate`
`make db_test_e2e_populate`

Ensure no errors and check for indexes on table columns.

## Code Review Verification Steps

* Any new migrations/schema changes:
  * [ ] Follow our guidelines for zero-downtime deploys (see [Zero-Downtime Deploys](./docs/database.md#zero-downtime-migrations))
  * [ ] Have been communicated to #dp3-engineering
  * [ ] Secure migrations have been tested using `bin/run-prod-migrations`
* [ ] Request review from a member of a different team.
* [ ] Have the Pivotal acceptance criteria been met for this change?

## References

* [Pivotal story](https://www.pivotaltracker.com/story/show/164870225) for this change
* [this article](https://sqlperformance.com/2012/11/t-sql-queries/benefits-indexing-foreign-keys) explains more about the approach used.

chrisgilmerproj commented on migrations/20190306205302_add-indexes-for-foreign-keys.up.fizz:
```diff
@@ -0,0 +1,92 @@
+
+add_index("shipments", "status", {});
+add_index("shipments", "actual_pickup_date", {});
+add_index("shipments", "actual_delivery_date", {});
```
Can you sort this file lexicographically so that similar models are stacked together?  It would make it much easier to read.  For instance, these `shipments` indexes ought to go above with the other `shipments` indexes. And having the columns sorted would be helpful (unless that causes some kind of run-time error).

chrisgilmerproj commented on migrations/20190306205302_add-indexes-for-foreign-keys.up.fizz:
```diff
@@ -0,0 +1,92 @@
+add_index("blackout_dates", "transportation_service_provider_id", {});
```
So this would be separate from the `blackout_dates_index`? The existing one covers a combination of:
- transportation_service_provider_id
- start_blackout_date
- end_blackout_date
- market
- source_gbloc

Are you just trying to target foreign keys on their own or do you want to capture the most common query pattern?

lynzt replied:
for now, I was adding according to foreign keys, but I agree, I'd like to add them by query pattern eventually. This was a first pass at trying to make things faster. At some point, seeing what keys are actually hit and how they're used would be great... (I'm sure some queries would do well w/ composite indexes)

lynzt replied:
Also, I see that I could drop the `transportation_service_provider_id` since, as you mention there's already a composite index w/ the first column being `transportation_service_provider_id`. 

An index on `transportation_service_provider_id` only isn't needed as it's redundant as it can use the existing index. (yes?)

chrisgilmerproj replied:
My experience is no, composite indexes will only be used if you're also querying for the other things as well. But I'd be happy to be proven wrong there.  I'm also not against being completionist here just so we are explicit that we did the work and others don't come along with questions later.  You can do some research but it's probably fine to have both.

chrisgilmerproj commented on migrations/20190306205302_add-indexes-for-foreign-keys.up.fizz:
```diff
@@ -0,0 +1,92 @@
+add_index("duty_stations", "address_id", {});
+add_index("duty_stations", "transportation_office_id", {});
+
+
```
You might want to add

```
add_index("electronic_orders_revisions", "electronic_order_id", {});
```

I'm guessing a more likely target will be this though:

```
add_index("electronic_orders_revisions", ["electronic_order_id", "seq_num"], {});
```

chrisgilmerproj commented on migrations/20190306205302_add-indexes-for-foreign-keys.up.fizz:
```diff
@@ -0,0 +1,92 @@
+
+add_index("move_documents", "document_id", {});
+add_index("move_documents", "personally_procured_move_id", {});
+add_index("move_documents", "shipment_id", {});
```
If you're just targeting foreign keys explicitly then I'd add:

```
add_index("move_documents", "move_id", {});
```

It already has one `move_id` and `document_id` which is probably the more common query.

lynzt replied:
I was targeting foreign keys that also didn't have a usable index...

The existing `move_documents_move_id_document_id_idx` can be used by `move_id` alone

chrisgilmerproj commented on migrations/20190306205302_add-indexes-for-foreign-keys.up.fizz:
```diff
@@ -0,0 +1,92 @@
+add_index("distance_calculations", "destination_address_id", {});
+
+add_index("documents", "service_member_id", {});
+
```
Let's add one for DPS:

```
add_index("dps_users", "login_gov_email", {});
```

chrisgilmerproj commented on migrations/20190306205302_add-indexes-for-foreign-keys.up.fizz:
```diff
@@ -0,0 +1,92 @@
+add_index("shipments", "destination_address_on_acceptance_id", {});
+
+add_index("signed_certifications", "submitting_user_id", {});
+add_index("signed_certifications", "move_id", {});
```
This table already has an index that combines `traffic_distribution_list_id`, `actual_pickup_date`, and `actual_delivery_date` but it could use:

```
add_index("signed_certifications", "traffic_distribution_list_id", {});
add_index("signed_certifications", "pickup_address_id", {});
add_index("signed_certifications", "secondary_pickup_address_id", {});
add_index("signed_certifications", "delivery_address_id", {});
add_index("signed_certifications", "partial_sit_delivery_address_id", {});
add_index("signed_certifications", "service_member_id", {});
add_index("signed_certifications", "shipping_distance_id", {});
add_index("signed_certifications", "shipping_address_on_acceptance_id", {});
```

lynzt replied:
adding the indexes you mention... 
Wondering about the `traffic_distribution_list_id` since we can use that already w/ the `shipments_index`?

chrisgilmerproj replied:
Something was up with my sql editor when I was looking at this earlier. I think it showed me the structure for `shipments` but the label was `signed_certifications`.  So can you double check I haven't screwed up with this recommendation?

lynzt replied:
I think you might be right... this looks like `shipments`...

chrisgilmerproj commented on migrations/20190306205302_add-indexes-for-foreign-keys.up.fizz:
```diff
@@ -0,0 +1,92 @@
+add_index("tsp_users", "user_id", {});
+add_index("tsp_users", "transportation_service_provider_id", {});
+
+add_index("uploads", "uploader_id", {});
```
Also add 

```
add_index("uploads", "document_id", {});
```

chrisgilmerproj commented on migrations/20190306205302_add-indexes-for-foreign-keys.up.fizz:
```diff
@@ -0,0 +1,92 @@
+add_index("uploads", "uploader_id", {});
+
+
+
```
I'd like us to add another one here for `users`:

```
add_index("users", "login_gov_email", {});
```

lynzt commented:
Added the additional indexes you requested...
Added indexes on any email fields I saw...
I did not add single field indexes on tables that had a composite index as we can use the existing index.

chrisgilmerproj commented:
When you're ready to try this on experimental let me know and I'm happy to walk you through the process.
